File: Stack/java/src/StringReverser.java
import java.util.Stack;

public class StringReverser {
    public String reverse(String input) {
//        Edge case test
        if (input == null)
            throw new IllegalArgumentException();

        Stack<Character> stack = new Stack<>();

        for (char ch : input.toCharArray())
            stack.push(ch);

        StringBuffer reversed = new StringBuffer();

        while (!stack.empty())
            reversed.append(stack.pop());

        return reversed.toString();
    }

}

File: 03-Stacks/cpp/StringReverser.cpp
#include "StringReverser.h"
#include <stack>

using namespace std;

string StringReverser::reverse(string input) {
    if (input.empty())
        throw "input is empty";

    stack<char> buf;

    for (auto ch: input) {
        buf.push(ch);
    }

    string reversed;

    while (!buf.empty()) {
        reversed += buf.top();
        buf.pop();
    }

    return reversed;
}
